File: gn-farm-go-server/internal/controller/health/health.controller.go
//####################################################################
// HEALTH CHECK CONTROLLER - CONTROLLER KIỂM TRA SỨC KHỊE HỆ THỐNG
// Package này chứa các controller kiểm tra sức khỏe và trạng thái của hệ thống
// 
// Chức năng chính:
// - Health Check: Kiểm tra tổng thể trạng thái hệ thống
// - Readiness Check: Kiểm tra sẵn sàng phục vụ requests 
// - Liveness Check: Kiểm tra ứng dụng còn hoạt động
// - Database Health: Kiểm tra kết nối PostgreSQL
// - Redis Health: Kiểm tra kết nối Redis cache
//
// Sử dụng cho:
// - Container orchestration (Kubernetes health probes)
// - Load balancer health checks
// - Monitoring và alerting systems
// - DevOps và infrastructure monitoring
// - API status dashboard
//
// Endpoints:
// - GET /health: Tổng quan trạng thái hệ thống
// - GET /ready: Kiểm tra sẵn sàng nhận traffic
// - GET /live: Kiểm tra ứng dụng còn sống
//
// Tác giả: GN Farm Development Team
// Phiên bản: 1.0
//####################################################################

// Package health chứa các controller kiểm tra sức khỏe và trạng thái của hệ thống
// Sử dụng cho monitoring, load balancer health checks, và container orchestration
package health

import (
	"context"  // Context cho các operations có timeout
	"net/http" // HTTP status codes
	"time"     // Time operations và duration tracking

	"gn-farm-go-server/global"       // Global variables (database, redis connections)
	"gn-farm-go-server/pkg/response" // Response utilities

	"github.com/gin-gonic/gin" // Gin web framework
)

// Health - Global instance của health controller (singleton pattern)
var Health = new(healthController)

// healthController - Controller xử lý các endpoint liên quan đến sức khỏe hệ thống
type healthController struct{}

// HealthStatus - Struct đại diện cho trạng thái tổng thể của hệ thống
// Chứa thông tin chi tiết về tất cả các dịch vụ và kết nối
type HealthStatus struct {
	Status    string                 `json:"status"`    // Trạng thái tổng thể ("healthy" hoặc "unhealthy")
	Timestamp time.Time              `json:"timestamp"` // Thời điểm kiểm tra
	Version   string                 `json:"version"`   // Phiên bản ứng dụng
	Services  map[string]ServiceInfo `json:"services"`  // Trạng thái của từng dịch vụ (database, redis, etc.)
	Uptime    string                 `json:"uptime"`    // Thời gian hoạt động liên tục
}

// ServiceInfo - Struct chứa thông tin chi tiết của một dịch vụ cụ thể
// Bao gồm trạng thái, thời gian phản hồi, lỗi (nếu có), và chi tiết bổ sung
type ServiceInfo struct {
	Status      string        `json:"status"`        // Trạng thái ("healthy" hoặc "unhealthy")
	ResponseTime string       `json:"response_time"` // Thời gian phản hồi
	Error       string        `json:"error,omitempty"` // Mô tả lỗi nếu có
	Details     interface{}   `json:"details,omitempty"` // Thông tin chi tiết bổ sung (stats, metrics)
}

// startTime - Thời điểm khởi động server, dùng để tính uptime
var startTime = time.Now()

// HealthCheck - Thực hiện kiểm tra sức khỏe tổng thể của hệ thống
// Chức năng:
// - Kiểm tra kết nối database (PostgreSQL)
// - Kiểm tra kết nối Redis
// - Tính toán uptime và trạng thái tổng quát
// - Trả về báo cáo chi tiết về tất cả các dịch vụ
// @Summary      Health check endpoint - Endpoint kiểm tra sức khỏe
// @Description  Trả về trạng thái sức khỏe của dịch vụ và các dependency
// @Tags         health
// @Accept       json
// @Produce      json
// @Success      200  {object}  HealthStatus
// @Failure      503  {object}  HealthStatus
// @Router       /health [get]
func (h *healthController) HealthCheck(ctx *gin.Context) {
	// Khởi tạo struct chứa thông tin sức khỏe ban đầu
	healthStatus := HealthStatus{
		Status:    "healthy",                    // Mặc định là healthy
		Timestamp: time.Now(),                   // Thời điểm hiện tại
		Version:   "1.0.0",                     // Phiên bản ứng dụng
		Services:  make(map[string]ServiceInfo), // Khởi tạo map chứa trạng thái các dịch vụ
		Uptime:    time.Since(startTime).String(), // Tính uptime từ lúc khởi động
	}

	// Kiểm tra kết nối database và lưu kết quả
	dbStatus := h.checkDatabase()
	healthStatus.Services["database"] = dbStatus

	// Kiểm tra kết nối Redis và lưu kết quả
	redisStatus := h.checkRedis()
	healthStatus.Services["redis"] = redisStatus

	// Xác định trạng thái tổng quát dựa trên trạng thái của tất cả dịch vụ
	overallStatus := "healthy"           // Mặc định là healthy
	statusCode := http.StatusOK          // HTTP 200 OK

	// Duyệt qua tất cả dịch vụ để kiểm tra trạng thái
	for _, service := range healthStatus.Services {
		if service.Status != "healthy" {
			overallStatus = "unhealthy"                    // Nếu có 1 dịch vụ unhealthy -> tổng thể unhealthy
			statusCode = http.StatusServiceUnavailable     // HTTP 503 Service Unavailable
			break                                          // Thoát khỏi vòng lặp
		}
	}

	// Cập nhật trạng thái tổng quát
	healthStatus.Status = overallStatus

	// Trả về JSON response với status code tương ứng
	ctx.JSON(statusCode, healthStatus)
}

// ReadinessCheck - Kiểm tra xem dịch vụ có sẵn sàng phục vụ request không
// Chức năng:
// - Kiểm tra các dependency quan trọng (database, redis)
// - Chỉ trả về ready khi tất cả dependency hoạt động bình thường
// - Sử dụng cho Kubernetes readiness probe
// @Summary      Readiness check endpoint - Endpoint kiểm tra sẵn sàng
// @Description  Trả về xem dịch vụ có sẵn sàng phục vụ requests không
// @Tags         health
// @Accept       json
// @Produce      json
// @Success      200  {object}  response.ResponseData
// @Failure      503  {object}  response.ErrorResponseData
// @Router       /ready [get]
func (h *healthController) ReadinessCheck(ctx *gin.Context) {
	// Kiểm tra các dependency quan trọng
	dbStatus := h.checkDatabase()   // Kiểm tra database
	redisStatus := h.checkRedis()   // Kiểm tra Redis

	// Chỉ trả về ready khi cả database và redis đều hoạt động tốt
	if dbStatus.Status == "healthy" && redisStatus.Status == "healthy" {
		// Dịch vụ sẵn sàng phục vụ
		response.SuccessResponse(ctx, response.ErrCodeSuccess, gin.H{
			"status": "ready",
			"timestamp": time.Now(),
		})
	} else {
		// Dịch vụ chưa sẵn sàng do có dependency bị lỗi
		response.ErrorResponse(ctx, response.ErrCodeInternalServerError, "Service not ready")
	}
}

// LivenessCheck - Kiểm tra xem dịch vụ có đang hoạt động không
// Chức năng:
// - Kiểm tra đơn giản xem ứng dụng có responsive không
// - Không phụ thuộc vào các dịch vụ bên ngoài
// - Sử dụng cho Kubernetes liveness probe
// - Luôn trả về success nếu process vẫn đang chạy
// @Summary      Liveness check endpoint - Endpoint kiểm tra sự sống
// @Description  Trả về xem dịch vụ có đang hoạt động không
// @Tags         health
// @Accept       json
// @Produce      json
// @Success      200  {object}  response.ResponseData
// @Router       /live [get]
func (h *healthController) LivenessCheck(ctx *gin.Context) {
	// Trả về thông tin cơ bản về trạng thái sống của ứng dụng
	response.SuccessResponse(ctx, response.ErrCodeSuccess, gin.H{
		"status": "alive",                           // Ứng dụng đang sống
		"timestamp": time.Now(),                     // Thời điểm kiểm tra
		"uptime": time.Since(startTime).String(),    // Thời gian hoạt động liên tục
	})
}

// checkDatabase - Kiểm tra kết nối và trạng thái của PostgreSQL database
// Chức năng:
// - Kiểm tra xem connection pool có được khởi tạo không
// - Thực hiện ping để kiểm tra kết nối thực tế
// - Lấy thống kê connection pool (active, idle connections)
// - Đo thời gian phản hồi
// @return ServiceInfo - Thông tin chi tiết về trạng thái database
func (h *healthController) checkDatabase() ServiceInfo {
	start := time.Now() // Bắt đầu đo thời gian

	// Kiểm tra xem database connection có được khởi tạo không
	if global.Pgdbc == nil {
		return ServiceInfo{
			Status:       "unhealthy",                            // Trạng thái không tốt
			ResponseTime: time.Since(start).String(),             // Thời gian xử lý
			Error:        "Database connection not initialized",  // Lỗi: chưa khởi tạo kết nối
		}
	}

	// Thực hiện ping để kiểm tra kết nối thực tế
	err := global.Pgdbc.Ping()
	responseTime := time.Since(start).String() // Tính thời gian phản hồi

	// Nếu ping thất bại
	if err != nil {
		return ServiceInfo{
			Status:       "unhealthy",   // Trạng thái không tốt
			ResponseTime: responseTime,   // Thời gian phản hồi
			Error:        err.Error(),   // Chi tiết lỗi
		}
	}

	// Lấy thống kê connection pool để monitor hiệu suất
	stats := global.Pgdbc.Stats()

	// Trả về thông tin trạng thái tốt kèm chi tiết
	return ServiceInfo{
		Status:       "healthy",     // Trạng thái tốt
		ResponseTime: responseTime,   // Thời gian phản hồi
		Details: map[string]interface{}{
			"open_connections": stats.OpenConnections, // Số kết nối đang mở
			"in_use":          stats.InUse,          // Số kết nối đang sử dụng
			"idle":            stats.Idle,           // Số kết nối đang idle
		},
	}
}

// checkRedis - Kiểm tra kết nối và trạng thái của Redis server
// Chức năng:
// - Kiểm tra xem Redis client có được khởi tạo không
// - Thực hiện PING command để kiểm tra kết nối thực tế
// - Lấy thống kê connection pool (hits, misses, timeouts)
// - Đo thời gian phản hồi
// @return ServiceInfo - Thông tin chi tiết về trạng thái Redis
func (h *healthController) checkRedis() ServiceInfo {
	start := time.Now() // Bắt đầu đo thời gian

	// Kiểm tra xem Redis client có được khởi tạo không
	if global.Rdb == nil {
		return ServiceInfo{
			Status:       "unhealthy",                        // Trạng thái không tốt
			ResponseTime: time.Since(start).String(),         // Thời gian xử lý
			Error:        "Redis connection not initialized", // Lỗi: chưa khởi tạo kết nối
		}
	}

	// Thực hiện PING command với context để kiểm tra kết nối
	ctx := context.Background()
	pong, err := global.Rdb.Ping(ctx).Result()
	responseTime := time.Since(start).String() // Tính thời gian phản hồi

	// Nếu PING thất bại
	if err != nil {
		return ServiceInfo{
			Status:       "unhealthy",   // Trạng thái không tốt
			ResponseTime: responseTime,   // Thời gian phản hồi
			Error:        err.Error(),   // Chi tiết lỗi
		}
	}

	// Lấy thống kê connection pool để monitor hiệu suất
	poolStats := global.Rdb.PoolStats()

	// Trả về thông tin trạng thái tốt kèm chi tiết
	return ServiceInfo{
		Status:       "healthy",     // Trạng thái tốt
		ResponseTime: responseTime,   // Thời gian phản hồi
		Details: map[string]interface{}{
			"ping_response": pong,        // Kết quả của PING command (thường là "PONG")
			"pool_stats": map[string]interface{}{
				"hits":        poolStats.Hits,        // Số lần sử dụng connection từ pool thành công
				"misses":      poolStats.Misses,      // Số lần phải tạo connection mới
				"timeouts":    poolStats.Timeouts,    // Số lần timeout khi lấy connection
				"total_conns": poolStats.TotalConns,  // Tổng số connection
				"idle_conns":  poolStats.IdleConns,   // Số connection đang idle
			},
		},
	}
}

File: gn-chat-service/src/controllers/health.controller.js
// Controller kiểm tra sức khỏe hệ thống và các dịch vụ liên kết
const mongoose = require("mongoose") // MongoDB ODM
const { Pool } = require("pg") // PostgreSQL pool
const redis = require("../config/redis") // Redis client

/**
 * Class quản lý các endpoint kiểm tra sức khỏe của ứng dụng
 * Cung cấp các endpoint:
 * - Health check: Kiểm tra tổng thể sức khỏe hệ thống
 * - Readiness check: Kiểm tra ứng dụng sẵn sàng phục vụ
 * - Liveness check: Kiểm tra ứng dụng còn hoạt động
 */

class HealthController {
  constructor() {
    this.startTime = Date.now() // Lưu thời điểm khởi động để tính uptime
  }

  /**
   * Kiểm tra sức khỏe tổng thể của hệ thống
   * Kiểm tra tất cả các dịch vụ: MongoDB, PostgreSQL, Redis
   * Trả về thông tin chi tiết về trạng thái của từng dịch vụ
   * @param {Object} req - HTTP request
   * @param {Object} res - HTTP response
   */
  async healthCheck(req, res) {
    try {
      // Khởi tạo thông tin sức khỏe cơ bản
      const healthStatus = {
        status: "healthy", // Trạng thái tổng thể
        timestamp: new Date().toISOString(), // Thời điểm kiểm tra
        version: process.env.npm_package_version || "1.0.0", // Phiên bản ứng dụng
        uptime: this.getUptime(), // Thời gian hoạt động
        services: {}, // Trạng thái các dịch vụ
      }

      // Kiểm tra tất cả các dịch vụ song song để tối ưu thời gian
      const [mongoStatus, postgresStatus, redisStatus] =
        await Promise.allSettled([
          this.checkMongoDB(), // Kiểm tra MongoDB
          this.checkPostgreSQL(), // Kiểm tra PostgreSQL
          this.checkRedis(), // Kiểm tra Redis
        ])

      // Xử lý kết quả kiểm tra của từng dịch vụ
      healthStatus.services.mongodb = this.getServiceResult(mongoStatus)
      healthStatus.services.postgresql = this.getServiceResult(postgresStatus)
      healthStatus.services.redis = this.getServiceResult(redisStatus)

      // Xác định trạng thái tổng thể dựa trên trạng thái của tất cả dịch vụ
      const allHealthy = Object.values(healthStatus.services).every(
        (service) => service.status === "healthy"
      )

      healthStatus.status = allHealthy ? "healthy" : "unhealthy"

      // Trả về HTTP status code phù hợp
      const statusCode = allHealthy ? 200 : 503
      res.status(statusCode).json(healthStatus)
    } catch (error) {
      // Xử lý lỗi không mong muốn
      res.status(500).json({
        status: "error",
        timestamp: new Date().toISOString(),
        error: error.message,
      })
    }
  }

  /**
   * Kiểm tra ứng dụng sẵn sàng phục vụ request
   * Chỉ trả về ready khi tất cả các dịch vụ cần thiết hoạt động bình thường
   */
  async readinessCheck(req, res) {
    try {
      const [mongoStatus, postgresStatus, redisStatus] =
        await Promise.allSettled([
          this.checkMongoDB(),
          this.checkPostgreSQL(),
          this.checkRedis(),
        ])

      const mongoHealthy =
        mongoStatus.status === "fulfilled" &&
        mongoStatus.value.status === "healthy"
      const postgresHealthy =
        postgresStatus.status === "fulfilled" &&
        postgresStatus.value.status === "healthy"
      const redisHealthy =
        redisStatus.status === "fulfilled" &&
        redisStatus.value.status === "healthy"

      if (mongoHealthy && postgresHealthy && redisHealthy) {
        res.status(200).json({
          status: "ready",
          timestamp: new Date().toISOString(),
        })
      } else {
        res.status(503).json({
          status: "not ready",
          timestamp: new Date().toISOString(),
          services: {
            mongodb: mongoHealthy,
            postgresql: postgresHealthy,
            redis: redisHealthy,
          },
        })
      }
    } catch (error) {
      res.status(503).json({
        status: "not ready",
        timestamp: new Date().toISOString(),
        error: error.message,
      })
    }
  }

  /**
   * Kiểm tra ứng dụng còn sống (liveness check)
   * Endpoint đơn giản để kiểm tra process còn hoạt động
   */
  async livenessCheck(req, res) {
    res.status(200).json({
      status: "alive",
      timestamp: new Date().toISOString(),
      uptime: this.getUptime(),
      pid: process.pid,
      memory: process.memoryUsage(),
      nodeVersion: process.version,
    })
  }

  /**
   * Kiểm tra kết nối MongoDB
   * Thực hiện ping để kiểm tra kết nối và đo response time
   */
  async checkMongoDB() {
    const start = Date.now()

    try {
      if (mongoose.connection.readyState !== 1) {
        throw new Error("MongoDB not connected")
      }

      // Perform a simple operation to test connection
      await mongoose.connection.db.admin().ping()

      const responseTime = Date.now() - start

      return {
        status: "healthy",
        responseTime: `${responseTime}ms`,
        details: {
          readyState: mongoose.connection.readyState,
          host: mongoose.connection.host,
          port: mongoose.connection.port,
          name: mongoose.connection.name,
        },
      }
    } catch (error) {
      return {
        status: "unhealthy",
        responseTime: `${Date.now() - start}ms`,
        error: error.message,
      }
    }
  }

  /**
   * Kiểm tra kết nối PostgreSQL
   * Tạo kết nối mới và thực hiện query test
   */
  async checkPostgreSQL() {
    const start = Date.now()

    try {
      const pool = new Pool({
        host: process.env.POSTGRES_HOST,
        port: process.env.POSTGRES_PORT,
        user: process.env.POSTGRES_USER,
        password: process.env.POSTGRES_PASSWORD,
        database: process.env.POSTGRES_DB,
        max: 1, // Only one connection for health check
        idleTimeoutMillis: 1000,
        connectionTimeoutMillis: 2000,
      })

      const client = await pool.connect()
      const result = await client.query("SELECT NOW()")
      client.release()
      await pool.end()

      const responseTime = Date.now() - start

      return {
        status: "healthy",
        responseTime: `${responseTime}ms`,
        details: {
          serverTime: result.rows[0].now,
        },
      }
    } catch (error) {
      return {
        status: "unhealthy",
        responseTime: `${Date.now() - start}ms`,
        error: error.message,
      }
    }
  }

  /**
   * Kiểm tra kết nối Redis
   * Sử dụng Redis client đã có để thực hiện ping
   */
  async checkRedis() {
    const start = Date.now()

    try {
      // Lấy Redis client từ module redis
      const redisClient = redis.getRedisClient() // Sửa tên hàm từ getClient thành getRedisClient

      if (!redisClient || !redisClient.isOpen) {
        throw new Error("Redis client not connected")
      }

      await redisClient.ping()

      const responseTime = Date.now() - start

      return {
        status: "healthy",
        responseTime: `${responseTime}ms`,
        details: {
          connected: redisClient.isOpen,
          ready: redisClient.isReady,
        },
      }
    } catch (error) {
      return {
        status: "unhealthy",
        responseTime: `${Date.now() - start}ms`,
        error: error.message,
      }
    }
  }

  // Xử lý kết quả từ Promise.allSettled
  getServiceResult(settledResult) {
    if (settledResult.status === "fulfilled") {
      return settledResult.value
    } else {
      return {
        status: "unhealthy",
        error: settledResult.reason.message,
      }
    }
  }

  // Tính toán và định dạng uptime thành chuỗi dễ đọc
  getUptime() {
    const uptimeMs = Date.now() - this.startTime
    const uptimeSeconds = Math.floor(uptimeMs / 1000)
    const hours = Math.floor(uptimeSeconds / 3600)
    const minutes = Math.floor((uptimeSeconds % 3600) / 60)
    const seconds = uptimeSeconds % 60

    return `${hours}h ${minutes}m ${seconds}s`
  }
}

// Export instance duy nhất của HealthController (Singleton pattern)
module.exports = new HealthController()
